doc/storage/api/api.rst

```diff
@@ -649,17 +680,144 @@ These definitions must be defined in the header file :file:`psa/protected_storag
 
     * ``data_offset + data_length <= capacity``
-
+    
     * Even if all parameters are correct, the function can fail in the case of a storage failure.
 
     On Success:
 
     * ``size = max(size, data_offset + data_length)``
 
     * ``capacity`` unchanged.
+    
+    * Data in the ranges 0 to ``data_offset`` is not modified. 
+    
+    * If ``data_offset + data_length < size`` then data in the range ``data_offset + data_length`` to ``size`` is not modified. 
+
+
+
+.. function:: psa_ps_rename
+
+   .. summary::
+        Atomically renames the storage location with the specified ``uid`` to a ``uid_new``.
+
+   .. param:: psa_storage_uid_t uid
+        The current identifier for the data.
+
+   .. param:: psa_storage_uid_t uid_new
+        The new identifier for the data.
+
+   .. param:: psa_storage_rename_flags_t rename_flags
+        The flags must be either ``PSA_STORAGE_FLAG_NONE`` or ``PSA_STORAGE_FLAG_OVERWRITE``
+
+   .. return:: psa_status_t
+        A status indicating the success or failure of the operation.
+
+   .. retval:: PSA_SUCCESS
+        The operation completed successfully.
+
+   .. retval:: PSA_ERROR_ALREADY_EXISTS
+        Storage with the specified ``uid_new`` already exists. 
+
+   .. retval:: PSA_ERROR_GENERIC_ERROR
+        The operation failed because of an unspecified internal failure.
+
+   .. retval:: PSA_ERROR_INVALID_ARGUMENT
+        The operation failed because either:
+
+        * ``uid`` is ``0``.
+        * ``uid_new`` is ``0``
+        * the ``psa_storage_rename_flags_t`` has a value set other than ``PSA_STORAGE_FLAG_OVERWRITE``
+
+   .. retval:: PSA_ERROR_NOT_PERMITTED
+        The operation failed because either:                
+        
+        * ``uid_new`` exists and was created with ``PSA_STORAGE_FLAG_WRITE_ONCE``.        
+        * ``uid_new`` exists and ``PSA_STORAGE_FLAG_OVERWRITE`` is not specified
+
+        * ``uid`` was created with ``PSA_STORAGE_FLAG_NO_CONFIDENTIALITY`` and ``uid_new`` exists and does not have that flag set
+        * ``uid`` was created with ``PSA_STORAGE_FLAG_NO_REPLAY_PROTECTION`` and ``uid_new`` exists and does not have that flag set
+
+   .. retval:: PSA_ERROR_NOT_SUPPORTED
+        The implementation does not support the operation.
+
+   .. retval:: PSA_ERROR_STORAGE_FAILURE
+        The operation failed because the physical storage has failed (Fatal error).     
+
+
+   The function renames ``uid`` to ``uid_new`` retaining the storage flags  that ``uid`` was created with. 
+
+   If the caller specifies ``PSA_STORAGE_FLAG_OVERWRITE`` the operation atomically replaces the exisitng contents of ``uid_new`` with those of ``uid``.
+
+   However, ``uid`` must not have been created with more permissive Storage flags than ``uid_new``
+        
+   Except in the case of ``PSA_ERROR_STORAGE_FAILURE``, in which case no guarantees can be made, the operation shall either succeed or leave storage unchanged. 
+
+
+
+
+.. function:: psa_ps_iterator_init
```

Comment:
The iteration API needs a 'usage' example to show how applications are expected to use the interfaces, and what kinds of operation can be done while iterating.

It must also be very clearly documented that there is no 'close'/'resource release' function for iterators
I think this aspect of the API design should be much more clearly stated, as it has a critical effect on the way applications use them, and on the possible implementation choices. In particular, a RoT service implementation cannot afford to have service-side state relating to an iterator, given that:

* An application is permitted to have more than one active iterator
* It is not assumed or required that the application always operates an iterator (calls 'next' on it) until it reaches the end of the iteration

I think this interface design is acceptable, and implementations can maintain adequate iterator state in the iterator context (in the application) to meet the requirements of the use cases for the API.